File: 2024/typescript/medium/topKFrequent.ts
function topKFrequent(nums: number[], k: number): number[] {
  if (nums.length === 1) return nums;
  let result: number[] = [];
  let count = {};

  for (let num of nums) {
    if (count[num]) {
      count[num]++;
    } else {
      count[num] = 1;
    }
  }

  let entries: [string | number, number][] = Object.entries(count);
  let sorted = entries.sort((a, b) => b[1] - a[1]);

  for (let i = 0; i < sorted.length; i++) {
    if (result.length === k) break;

    result.push(Number(sorted[i][0]));
  }

  return result;
}

File: 2024/go/medium/topKFrequent.go
package medium

import (
	"slices"
	"sort"
)

//lint:ignore U1000 Ignore unused function
func topKFrequent(nums []int, k int) []int {
	if len(nums) == 1 {
		return nums
	}

	slices.Sort(nums)

	count := make(map[int]int)

	for _, v := range nums {
		if count[v] != 0 {
			count[v]++
		} else {
			count[v] = 1
		}
	}

	values := make([]int, 0, len(count))

	for k := range count {
		values = append(values, k)
	}

	sort.SliceStable(values, func(i, j int) bool {
		return count[values[i]] < count[values[j]]
	})

	var result []int

	for i := len(values) - 1; i >= 0; i-- {
		if len(result) != k {
			result = append(result, values[i])
		}
	}

	return result
}
